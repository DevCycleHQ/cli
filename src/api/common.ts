export const BASE_URL = 'https://api.devcycle.com/'
export const AUTH_URL = 'https://auth.devcycle.com/'

export const buildHeaders = (token: string) => ({
    'Content-Type': 'application/json',
    Authorization: token,
})
