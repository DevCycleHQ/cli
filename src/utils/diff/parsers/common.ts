import parse from 'parse-diff'
import { ParseOptions, VariableMatch } from './types'

export abstract class BaseParser {
    abstract identity: string
    clientNames: string[]
    abstract variableMethodPattern: RegExp
    abstract variableNameCapturePattern: RegExp

    constructor(protected options: ParseOptions) {
        this.clientNames = options.clientNames ?? ['dvcClient']
    }

    buildRegexPattern() {
        return new RegExp(
            new RegExp(`(?:${this.clientNames.join('|')})`).source
            + this.variableMethodPattern.source
            + this.variableNameCapturePattern.source
        )
    }

    formatMatch(name: string, file: parse.File, change: parse.Change): VariableMatch {
        return {
            name,
            fileName: file.to ?? '',
            line: change.type === 'normal' ? change.ln1 : change.ln,
            mode: change.type === 'add' ? 'add' : 'remove'
        }
    }

    abstract match(content: string, options: ParseOptions): string | null

    parse(file: parse.File, options: ParseOptions = {}): VariableMatch[] {
        const results: VariableMatch[] = []
        for (const chunk of file.chunks) {
            for (const change of chunk.changes) {
                const match = this.match(change.content, options)
                if (match) {
                    results.push(this.formatMatch(match, file, change))
                }
            }
        }
        return results
    }
}
