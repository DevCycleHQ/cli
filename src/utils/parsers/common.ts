import parse from 'parse-diff'
import { ParseOptions, VariableDiffMatch, Range, MultilineUsageChunk, VariableUsageMatch } from './types'
import * as usage from '../../commands/usages/types'

type MultilineDiffChunk = {
    content: string
    changes: ParsedChange[]
}

type MatchWithRange = {
    // Variable name
    name: string
    // Start/end indicies of matching substring
    range: Range
    // true if the found key is not wrapped in quotes
    isUnknown?: boolean
}

export type MatchResult = {
    // true if the found key is not wrapped in quotes
    isUnknown?: boolean
    // content matching the pattern
    content: string
    // variable key found within match
    name: string
    // index of matching substring within parent string
    index: number
}

class ParsedChange {
    type: string
    content: string
    ln: number
    range: Range = { start: 0, end: 0 }

    constructor(change: parse.Change) {
        this.type = change.type
        this.content = change.content
        this.ln = change.type === 'normal' ? change.ln2 : change.ln
    }

    parseContent() {
        return this.type === 'normal'
            ? this.content.trim()
            : this.content.slice(1).trim()
    }

    isComment(commentCharacters: string[]) {
        const changeContent = this.parseContent()
        return commentCharacters
            .map((char) => changeContent.indexOf(char) === 0)
            .some(Boolean)
    }

    format(type: 'add' | 'del'): parse.AddChange | parse.DeleteChange {
        const baseChange = {
            type,
            ln: this.ln,
            content: this.content
        }
        return type === 'add'
            ? { ...baseChange, type, add: true }
            : { ...baseChange, type, del: true }

    }
}

export abstract class BaseParser {
    abstract identity: string
    // whether to add the client name (eg. dvcClient) at the beginning of the pattern
    matchClientName = true

    // list of custom client names to also match against
    clientNames: string[]

    // regex pattern to use to match the start of the "variable" method call, eg. `.variable(`
    abstract variableMethodPattern: RegExp

    // a list of patterns to use to match each positional parameter expected in the variable call,
    // and the closing bracket, eg. user, variable, default)
    orderedParameterPatterns: RegExp[] | null = null

    // delimiter to use when detecting "named" parameters, eg. dvcClient.variable(variable=variable, default=default)
    namedParameterDelimiter = '='

    // a map of named parameter names to the patterns used to detect them eg. {user: /someregex/}
    namedParameterPatternMap: Record<string, RegExp> | null = null

    // comment characters to look for when matching. Varies by language.
    commentCharacters: string[] = ['//']

    constructor(extension: string, protected options: ParseOptions) {
        this.clientNames = [...(options.clientNames || []), 'dvcClient']
    }

    private buildParameterPattern(
        namedParameterPatternMap: Record<string, RegExp> | null,
        orderedParameterPatterns: RegExp[] | null
    ) {
        const namedParameters = namedParameterPatternMap
            ? Object
                .entries(namedParameterPatternMap)
                .map(([key, pattern]) =>
                    new RegExp(
                        `(?=.*?${key}${this.namedParameterDelimiter}\\s*${pattern.source})`
                    ).source
                )
                .join('')
                .concat(/[^)]*\)/.source)
            : null

        const orderedParameters = orderedParameterPatterns
            ? orderedParameterPatterns
                .map((p) => p.source)
                .join(/\s*,\s*/.source)
                .concat(/\)/.source)
            : null

        return orderedParameters && namedParameters
            ? new RegExp(`(?:(?:${orderedParameters})|(?:${namedParameters}))`).source
            : orderedParameters || namedParameters
    }

    private buildClientNamePattern() {
        return this.matchClientName
            ? new RegExp(`(?:${this.clientNames.join('|')})`).source
            : ''
    }

    private buildRegexPattern() {
        const clientNamePattern = this.buildClientNamePattern()

        const parameterPattern = this.buildParameterPattern(
            this.namedParameterPatternMap,
            this.orderedParameterPatterns
        )

        return new RegExp(
            clientNamePattern
            + this.variableMethodPattern.source
            + parameterPattern
        )
    }

    private extractUsageInformation(file: usage.File, match: MatchWithRange): MultilineUsageChunk {
        const lines: usage.LineItem[] = []
        let index = 0
    
        for (const line of file.lines) {
            const trimmedContent = line.content.trim()
            if (!trimmedContent.length) continue

            const { start: matchStartIndex, end: matchEndIndex } = match.range
            const lineStartIndex = index
            const lineEndIndex = lineStartIndex + trimmedContent.length - 1
            if (
                (matchStartIndex >= lineStartIndex && matchStartIndex <= lineEndIndex)
                || (lineStartIndex >= matchStartIndex && lineStartIndex <= matchEndIndex)
            ) {
                lines.push(line)
            }
            index = lineEndIndex + 1
        }

        return {
            content: lines.reduce((prev, curr) => prev + (prev ? '\n' : '') + curr.content, ''),
            start: lines.length ? lines[0].ln + 1 : -1,
            end: lines.length ? lines[lines.length - 1].ln + 1 : -1
        }
    }

    /**
     * Given a chunk from parse-diff, aggregate added and removed changes.
     * Each resulting chunk includes the multi-line content as a single string,
     * and an array of changes, each with a range object describing the start/end
     * indicies of the substring within the content string.
     */
    private aggregateMultilineChunks(chunk: parse.Chunk): { added: MultilineDiffChunk, removed: MultilineDiffChunk } {
        const added: MultilineDiffChunk = {
            content: '',
            changes: []
        }
        const removed: MultilineDiffChunk = {
            content: '',
            changes: []
        }

        const pushChangeToMultilineDiffChunk = (chunk: MultilineDiffChunk, change: ParsedChange) => {
            change.range.start = chunk.content.length

            const changeContent = change.parseContent()
            if (!changeContent.length) return

            chunk.content += changeContent
            change.range.end = chunk.content.length - 1

            chunk.changes.push(change)
        }

        for (const change of chunk.changes) {
            const parsedChange = new ParsedChange(change)
            if (parsedChange.isComment(this.commentCharacters)) continue

            if (parsedChange.type === 'add') {
                pushChangeToMultilineDiffChunk(added, parsedChange)
            } else if (parsedChange.type === 'del') {
                pushChangeToMultilineDiffChunk(removed, parsedChange)
            } else if (parsedChange.type === 'normal') {
                pushChangeToMultilineDiffChunk(added, parsedChange)
                pushChangeToMultilineDiffChunk(removed, parsedChange)
            }
        }
        return { added, removed }
    }

    printRegexPattern(): void {
        console.log(`Pattern for ${this.identity} parser: ${this.buildRegexPattern().source}`)
    }

    match(content: string): MatchResult | null {
        const pattern = this.buildRegexPattern()

        const matches = pattern.exec(content)
        if (matches) {
            const varName = (matches[1] || matches[2]).trim()
            return {
                isUnknown: !varName.match(/^["'].*["']$/),
                // position or named parameter match
                name: varName.replace(/^["']|["']$/g, ''),
                content: matches[0],
                index: matches.index
            }
        }

        return null
    }

    /**
     * Given a string, returns all matches found in the string.
     * Each match includes the variable name and the character range of the
     * matching substring.
     */
    private getAllMatches(matchString: string): MatchWithRange[] {
        const allMatches: MatchWithRange[] = []
        let cursorPosition = 0

        while (cursorPosition < matchString.length) {
            const substring = matchString.slice(cursorPosition)
            const match = this.match(substring)

            if (!match) break

            const { content, index, name, isUnknown } = match

            const startIndex = cursorPosition + index
            const endIndex = startIndex + content.length - 1

            allMatches.push({
                name,
                isUnknown,
                range: {
                    start: startIndex,
                    end: endIndex
                }
            })
            cursorPosition = endIndex + 1
        }

        return allMatches
    }

    private formatMatch(match: MatchWithRange, file: parse.File, change: parse.Change): VariableDiffMatch {
        return {
            name: match.name,
            fileName: file.to ?? '',
            line: change.type === 'normal' ? change.ln1 : change.ln,
            mode: change.type === 'add' ? 'add' : 'remove',
            ...(match.isUnknown ? { isUnknown: true } : {})
        }
    }

    /**
     * Given an array of matches, find the first change object corresponding to the match.
     * Also verify that the match is associated with at least one add/delete change object.
     */
    private formatMatches(file: parse.File, matches: MatchWithRange[], { changes }: MultilineDiffChunk) {
        const results: VariableDiffMatch[] = []

        matches.forEach((match) => {
            const { start: matchStartIndex, end: matchEndIndex } = match.range

            // Get all changes within the range of the match
            const matchingChanges = changes.filter(({ range }) => {
                const { start: changeStartIndex, end: changeEndIndex } = range
                return (matchStartIndex >= changeStartIndex && matchStartIndex <= changeEndIndex)
                    || (changeStartIndex >= matchStartIndex && changeStartIndex <= matchEndIndex)
            })

            // Remove "normal" change objects
            const validChange = matchingChanges.find(({ type }) => type !== 'normal')

            // If we have an add/del change, update the starting change to have that type and push to results
            if (validChange) {
                const formattedChange = matchingChanges[0].format(
                    validChange.type as 'add' | 'del'
                )
                results.push(this.formatMatch(match, file, formattedChange))
            }
        })

        return results
    }

    parseDiffs(file: parse.File): VariableDiffMatch[] {
        const results: VariableDiffMatch[] = []

        for (const chunk of file.chunks) {

            const { added: addedChunk, removed: removedChunk } = this.aggregateMultilineChunks(chunk)
            const addedMatches = this.getAllMatches(addedChunk.content)
            const removedMatches = this.getAllMatches(removedChunk.content)

            results.push(
                ...this.formatMatches(file, addedMatches, addedChunk),
                ...this.formatMatches(file, removedMatches, removedChunk)
            )
        }
        return results
    }

    parseFile(file: usage.File): VariableUsageMatch[] {
        const result: VariableUsageMatch[] = []
        let fileContent = ''
        for (const line of file.lines) {
            fileContent = fileContent.concat(line.content.trim()) 
        }

        const matches = this.getAllMatches(fileContent)

        for (const match of matches) {
            const usage = this.extractUsageInformation(file, match)
            if (usage.start !== -1 && usage.end !== -1) {
                result.push({
                    name: match.name,
                    line: usage.start,
                    fileName: file.name,
                    content: usage.content
                })
            }
        }

        return result
    }
}
