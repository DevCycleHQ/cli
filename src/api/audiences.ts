import apiClient from './apiClient'
import { buildHeaders } from './common'

const BASE_URL = '/v1/projects/:project/audiences'

export const fetchAudiences = async (
    token: string,
    project_id: string
) => {
    return apiClient.get(`${BASE_URL}`, {
        headers: buildHeaders(token),
        params: {
            project: project_id,
        }
    })
}