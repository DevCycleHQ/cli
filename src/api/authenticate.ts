import axios from 'axios'
import { AUTH_URL } from './common'

export const authenticate = async (client_id: string, client_secret: string): Promise<string> => {
    const url = new URL("/oauth/token", AUTH_URL)

    const response = await axios.post(url.href, {
        grant_type: "client_credentials",
        client_id,
        client_secret,
        audience: "https://api.devcycle.com",
    })

    return response.data.access_token
}
