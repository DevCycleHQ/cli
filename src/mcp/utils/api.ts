import { DevCycleAuth } from './auth'
import { fetchFeatures, createFeature } from '../../api/features'
import { fetchVariables } from '../../api/variables'
import { fetchEnvironmentByKey } from '../../api/environments'
import { enableTargeting, disableTargeting } from '../../api/targeting'

function getErrorMessage(error: unknown): string {
    if (error instanceof Error) {
        return error.message
    }
    return String(error)
}

function ensureError(error: unknown): Error {
    if (error instanceof Error) {
        return error
    }
    return new Error(String(error))
}

export class DevCycleApiClient {
    constructor(private auth: DevCycleAuth) {}

    async listFeatures(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error('MCP listFeatures params:', JSON.stringify(args, null, 2))

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const authToken = this.auth.getAuthToken()
            const projectKey = this.auth.getProjectKey()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchFeatures(authToken, projectKey, query)
            console.error(
                'MCP listFeatures result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listFeatures error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            console.error('Error during listFeatures:', getErrorMessage(error))
            throw ensureError(error)
        }
    }

    async listVariables(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error(
            'MCP listVariables params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchVariables(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                query,
            )
            console.error(
                'MCP listVariables result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listVariables error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async getSdkKeys(
        environmentKey: string,
        keyType?: 'mobile' | 'server' | 'client',
    ) {
        console.error(
            'MCP getSdkKeys params:',
            JSON.stringify({ environmentKey, keyType }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const environment = await fetchEnvironmentByKey(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                environmentKey,
            )

            const sdkKeys = environment.sdkKeys

            let result
            if (keyType) {
                result = {
                    [keyType]: sdkKeys[keyType],
                }
            } else {
                result = {
                    mobile: sdkKeys.mobile,
                    server: sdkKeys.server,
                    client: sdkKeys.client,
                }
            }

            console.error(
                'MCP getSdkKeys result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP getSdkKeys error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async enableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP enableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await enableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP enableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP enableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async disableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP disableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await disableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP disableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP disableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async createFeature(args: {
        key?: string
        name?: string
        description?: string
        type?: 'release' | 'experiment' | 'permission' | 'ops'
        interactive?: boolean
    }) {
        console.error(
            'MCP createFeature params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            if (args.interactive) {
                // For interactive mode, we would need to implement prompts
                // For now, return an error asking for explicit parameters
                throw new Error(
                    'Interactive mode not yet supported in MCP. Please provide explicit parameters: key, name, description, type',
                )
            }

            if (!args.key || !args.name) {
                throw new Error(
                    'Feature key and name are required when not using interactive mode',
                )
            }

            const featureData = {
                key: args.key,
                name: args.name,
                description: args.description || '',
                type: args.type || 'release',
            }

            const result = await createFeature(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureData,
            )
            console.error(
                'MCP createFeature result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP createFeature error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }
}
