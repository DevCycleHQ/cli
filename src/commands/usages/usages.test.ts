import { test } from '@oclif/test'
import { expect } from 'vitest'
import { BASE_URL } from '../../api/common'

describe('usages', () => {
    const projectKey = 'test-project'
    const mockVariable = {
        _id: '648a0d55c4e88cd4c4544c58',
        _project: '63b5ee5de6e91987bae47f01',
        _feature: '646f8bb69302b0862fd68a39',
        name: 'SPAM SPAM SPAM',
        key: 'variable-from-api',
        description: 'spammity spam ',
        type: 'String',
        status: 'active',
        source: 'cli',
        _createdBy: 'google-oauth2|111559006563333334214',
        createdAt: '2023-06-14T18:56:21.270Z',
        updatedAt: '2023-06-14T18:56:21.270Z',
    }

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/nodejs.js',
        ])
        .it('runs against a node test file', (ctx) => {
            expect(ctx.stdout).toMatchSnapshot()
        })

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/[org_id]/nodejs.js',
        ])
        .it(
            'runs against a node test file with special characters in the path',
            (ctx) => {
                expect(ctx.stdout).toMatchSnapshot()
            },
        )

    test.stdout()
        .command(['usages', '--include', 'test-utils/fixtures/usages/*/*.js'])
        .it('runs against all js files', (ctx) => {
            expect(ctx.stdout).toMatchSnapshot()
        })

    test.nock(BASE_URL, (api) =>
        api
            .get(
                `/v1/projects/${projectKey}/variables?perPage=1000&page=1&status=active`,
            )
            .reply(200, [mockVariable]),
    )
        .stdout()
        .command([
            'usages',
            '--only-unused',
            '--include',
            'test-utils/fixtures/usages/nodejs.js',
            '--project',
            projectKey,
        ])
        .it(
            'runs against a node test file and only returns variables not found as a variable in the api',
            (ctx) => {
                expect(ctx.stdout).toMatchSnapshot()
            },
        )

    test.stdout()
        .command(['usages', '--include', 'test-utils/fixtures/usages/react.js'])
        .it('runs against a react test file', (ctx) => {
            expect(ctx.stdout).toMatchSnapshot()
        })

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/[org_id]/react.js',
        ])
        .it(
            'runs against a react test file with special characters in the path',
            (ctx) => {
                expect(ctx.stdout).toMatchSnapshot()
            },
        )

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/golang.go',
        ])
        .it('runs against a go test file', (ctx) => {
            expect(ctx.stdout).toMatchSnapshot()
        })

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/[org_id]/golang.go',
        ])
        .it(
            'runs against a react test file with special characters in the path',
            (ctx) => {
                expect(ctx.stdout).toMatchSnapshot()
            },
        )

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/sample.dart',
        ])
        .it('runs against a dart test file', (ctx) => {
            expect(ctx.stdout).toMatchSnapshot()
        })

    test.stdout()
        .command([
            'usages',
            '--include',
            'test-utils/fixtures/usages/[org_id]/sample.dart',
        ])
        .it(
            'runs against a dart test file with special characters in the path',
            (ctx) => {
                expect(ctx.stdout).toMatchSnapshot()
            },
        )
})
