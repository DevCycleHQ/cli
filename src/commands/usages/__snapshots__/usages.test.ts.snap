// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usages > runs against a dart test file 1`] = `
"
DevCycle Variable Usage:

1. string-variable
	- test-utils/fixtures/usages/sample.dart:L81
2. boolean-variable
	- test-utils/fixtures/usages/sample.dart:L92
3. integer-variable
	- test-utils/fixtures/usages/sample.dart:L102
4. decimal-variable
	- test-utils/fixtures/usages/sample.dart:L112
5. json-array-variable
	- test-utils/fixtures/usages/sample.dart:L122
6. json-object-variable
	- test-utils/fixtures/usages/sample.dart:L135
"
`;

exports[`usages > runs against a dart test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. string-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L81
2. boolean-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L92
3. integer-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L102
4. decimal-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L112
5. json-array-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L122
6. json-object-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L135
"
`;

exports[`usages > runs against a go test file 1`] = `
"
DevCycle Variable Usage:

1. hello-test
	- test-utils/fixtures/usages/golang.go:L1
"
`;

exports[`usages > runs against a node test file 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/nodejs.js:L1
	- test-utils/fixtures/usages/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/nodejs.js:L37
"
`;

exports[`usages > runs against a node test file and only returns variables not found as a variable in the api 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/nodejs.js:L1
	- test-utils/fixtures/usages/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/nodejs.js:L25
8. multiline-extra-comma
	- test-utils/fixtures/usages/nodejs.js:L37
"
`;

exports[`usages > runs against a node test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L1
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L37
"
`;

exports[`usages > runs against a react test file 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/react.js:L1
	- test-utils/fixtures/usages/react.js:L6
	- test-utils/fixtures/usages/react.js:L11
"
`;

exports[`usages > runs against a react test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/react.js:L1
	- test-utils/fixtures/usages/[org_id]/react.js:L6
	- test-utils/fixtures/usages/[org_id]/react.js:L11
"
`;

exports[`usages > runs against a react test file with special characters in the path 2`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/golang.go:L1
"
`;

exports[`usages > runs against all js files 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L1
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L2
	- test-utils/fixtures/usages/[org_id]/react.js:L1
	- test-utils/fixtures/usages/[org_id]/react.js:L6
	- test-utils/fixtures/usages/[org_id]/react.js:L11
2. single-quotes
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L37
"
`;

exports[`usages runs against a dart test file 1`] = `
"
DevCycle Variable Usage:

1. string-variable
	- test-utils/fixtures/usages/sample.dart:L81
2. boolean-variable
	- test-utils/fixtures/usages/sample.dart:L92
3. integer-variable
	- test-utils/fixtures/usages/sample.dart:L102
4. decimal-variable
	- test-utils/fixtures/usages/sample.dart:L112
5. json-array-variable
	- test-utils/fixtures/usages/sample.dart:L122
6. json-object-variable
	- test-utils/fixtures/usages/sample.dart:L135
"
`;

exports[`usages runs against a dart test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. string-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L81
2. boolean-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L92
3. integer-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L102
4. decimal-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L112
5. json-array-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L122
6. json-object-variable
	- test-utils/fixtures/usages/[org_id]/sample.dart:L135
"
`;

exports[`usages runs against a go test file 1`] = `
"
DevCycle Variable Usage:

1. hello-test
	- test-utils/fixtures/usages/golang.go:L1
"
`;

exports[`usages runs against a node test file 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/nodejs.js:L1
	- test-utils/fixtures/usages/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/nodejs.js:L37
"
`;

exports[`usages runs against a node test file and only returns variables not found as a variable in the api 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/nodejs.js:L1
	- test-utils/fixtures/usages/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/nodejs.js:L25
8. multiline-extra-comma
	- test-utils/fixtures/usages/nodejs.js:L37
"
`;

exports[`usages runs against a node test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L1
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L2
2. single-quotes
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L37
"
`;

exports[`usages runs against a react test file 1`] = `
"
DevCycle Variable Usage:

1. simple-case
	- test-utils/fixtures/usages/react.js:L1
	- test-utils/fixtures/usages/react.js:L6
	- test-utils/fixtures/usages/react.js:L11
"
`;

exports[`usages runs against a react test file with special characters in the path 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/react.js:L1
	- test-utils/fixtures/usages/[org_id]/react.js:L6
	- test-utils/fixtures/usages/[org_id]/react.js:L11
"
`;

exports[`usages runs against a react test file with special characters in the path 2`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/golang.go:L1
"
`;

exports[`usages runs against all js files 1`] = `
"
DevCycle Variable Usage:

1. special-character-path
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L1
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L2
	- test-utils/fixtures/usages/[org_id]/react.js:L1
	- test-utils/fixtures/usages/[org_id]/react.js:L6
	- test-utils/fixtures/usages/[org_id]/react.js:L11
2. single-quotes
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L4
3. multi-line
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L10
4. multi-line-comment
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L20
5. user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L23
6. user-constructor
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L24
7. multi-line-user-object
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L25
8. variable-from-api
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L35
9. multiline-extra-comma
	- test-utils/fixtures/usages/[org_id]/nodejs.js:L37
"
`;
